### ISSUE TITLE
fatal error: 'stdio.h' file not found on Windows with binary release

### ISSUE THREAD

**Issue by randrew** [zdevito/terra#177]
Hi. I downloaded the binary release instead of building from source (as was recommended by the terra guide) on Windows. I downloaded `release-2016-03-25`. When I try any example that makes use of terralib.includec, terra seems to be unable to find any C library files unless I put them in the same directory as the .t file I'm trying to run. Obviously, that's a bad idea for things like stdio.h.

I saw in #92 and #120 that these system C files are supposed to somehow be included internally in the terra binary release, but it doesn't seem to work for me.

If I run the terra test suite (in share/terra/tests), 442 pass and 70 fail. The failing tests all appear to be ones which use terralib.includec.

**Comment by zdevito** [comment]
This happens when Visual Studio is installed in a different place from
where it was installed in the binary releases. The newest src releases have
a fix for it that uses the Window's registry to find where Visual Studio is
installed, and I will try to rebuild the binaries soon so that it can be
used.

For C headers, right now the only C files that are included internally are
the Clang-specific header files (like stddef.h) that individual compilers
provide. Things like 'stdio.h' still come from the libc installed on the
machine.

To work around this now, you can set the terralib.includepath, e.g.:

    terra.includepath = "/usr/local/include;..;"

Inside the application so that it searches more directories.

**Comment by randrew** [comment]
Thanks for the reply! Manually setting the includepath is letting me work with libc now.

**Comment by mxgbs** [comment]
Hi, where can I do the includepath? Visual studio? Or environment variable? Sorry. New to this? Moreover, What path should I include?

**Comment by elliottslaughter** [comment]
@mxgbs You can set the environment variable `INCLUDE_PATH` when you run Terra, or you can add a line to your Terra script to set `terralib.includepath` as shown above.

The exact path is going to depend on what Visual Studio you have installed, and unfortunately I don't have access to a machine with this installed so I can look myself. You might try looking here: https://stackoverflow.com/questions/13786112/where-can-i-find-the-implementation-of-stdio-h-in-visual-studio

Hope that helps.

**Comment by mxgbs** [comment]
Thank you. Actually I tried to include C:\Program Files (x86)\Microsoft
Visual Studio 14.0\VC to environment variable, however, it doesn't work. I
need to install opt based on terra. Thus, it would be better for me to find
a way to solve the problem globally, instead of defining in the script.

On Wed, Nov 29, 2017 at 4:11 PM, Elliott Slaughter <notifications@github.com

**Comment by mxgbs** [comment]
How can I add terralib.includepath to the script? Can you show me an
example? have no idea where to add it.

terralib.includepath = "/usr/local/include;..;"

local c = terralib.includec("stdio.h")

terra hello()
c.printf("hello, world\n")
end

terralib.saveobj("hello",{main = hello})
hello()

**Comment by elliottslaughter** [comment]
Yes, that's exactly right, you just have to modify the value you assign to `terralib.includepath` to point to do the directory that contains `stdio.h`. Note in your previous email you said you set the value to `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC` but if the link I posted above is right then it needs to be `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\crt\src`. At any rate, the directory you point it to needs to literally contain the file `stdio.h`. (I.e. you can't just point it to the top of the MSVC install, it's not that smart.)

**Comment by mxgbs** [comment]
I tried both crt\src and sdk, both not work independently or combined.
Google and use some other suggestions, it's still not work.

**Comment by elliottslaughter** [comment]
The problem has to be one of the following:

 1. `stdio.h` is not located in the directories you added to the include path.
 2. One of the transitive dependencies of `stdio.h` (i.e. via an `#include` directive inside `stdio.h`) is not on the path.
 3. Something is going wrong in `terralib.includec` or in your assignment of `terralib.includepath`.

I think we can rule out (3) by creating a dummy header file `my_header.h` (which can just be empty for this purpose), placing it in a directory somewhere, and testing the same program with `stdio.h` replaced by `my_header.h`. Note: for completeness you should make sure to set `terralib.includepath` to the absolute path to this file, and you should make sure the path has similar features (i.e. spaces and parens) to the other path you were using.

If that works, then you probably need to take a hard look at `stdio.h` and make sure (a) it's actually where you think it is, and (b) you include the paths of any dependencies it includes via `#include`.

**Comment by elliottslaughter** [comment]
Minor correction: when doing the test for (3) above, you need to set the path to the directory that contains the dummy header file, not the file itself.

**Comment by mxgbs** [comment]
Hi Elliott,

Thank you. I found out the right way to include that is as following(II).
However, new error popped up as in I. Do you have any clue how to solve
that? Besides, I also tried to add those two folder to the environment
variable, and use "SET" to check if those are correctly included. However,
even with them included in the environment variable, the test case is still
not work. The only way I found out to let it work is to include that in the
script.
I. terra\bin>terra ..\share\terra\tests\hello.t
llvm: program not executable

stack traceback:
        [C]: in function 'saveobj'
        C:\Users\zdevito\terra\src\terralib.lua:3979: in function 'saveobj'
        ..\share\terra\tests\hello.t:8: in main chunk


II. terralib.includepath = [[C:\Program Files (x86)\Windows
Kits\10\Include\10.0.16299.0\ucrt;C:\Program Files (x86)\Microsoft Visual
Studio 14.0\VC\include;C:\Users\dengya\Documents\Optlang\terra\bin]]
local c = terralib.includec("stdio.h")

terra hello()
c.printf("hello, world\n")
end

terralib.saveobj("hello",{main = hello})
hello()

On Thu, Nov 30, 2017 at 9:36 AM, Elliott Slaughter <notifications@github.com

**Comment by mxgbs** [comment]
Hi, @Elliott
Tried to modify terralib.saveobj("hello",{main = hello})
to  terralib.saveobj("hello.o",{main = hello}), it worked. hello.o
generated. But when generating executable file .o with g++ hello.o -o
hello, error popped as:.
hello.o:(.text[printf]+0x28): undefined reference to `__acrt_iob_func'
hello.o:(.text[printf]+0x42): undefined reference to
`__stdio_common_vfprintf'
collect2.exe: error: ld returned 1 exit status

Without change "hello" to "hello.o", the same error occurs as before. Then
I googled a little bit, it seems that some intrinsic functions are defined
differently in VS2015.
https://groups.google.com/forum/#!topic/dislin-users/kagVl9LVFJA . Do you
know how to solve that? Thank you.
llvm: program not executable

stack traceback:
        [C]: in function 'saveobj'
        C:\Users\zdevito\terra\src\terralib.lua:3979: in function 'saveobj'
        ..\share\terra\tests\hello.t:8: in main chunk

**Comment by mxgbs** [comment]
I think the general solution is here
https://community.microfocus.com/microfocus/mainframe_solutions/enterprise_developer_21/w/knowledge_base/27927/compatibility-issues-with-visual-studio-2015,
but I'm not sure how to apply to terra.

Sincerely,

Yan

**Comment by elliottslaughter** [comment]
I think if you want to use `saveobj` to generate a `.o` file you will need to use the MSVC compiler to link the final executable, given that `stdio.h` came from that compiler. (You mentioned using `g++` above.)

I unfortunately don't know anything about running MSVC, but from the article you linked it sounds like there is a legacy version of the functions you can link to if necessary? It's not completely clear if you need those or not.

**Comment by mxgbs** [comment]
No. I don't want to generate .o by saveobj. The problem is if I just try
saveobj to generate executable file, I will ge the error as I prviously
attached:

llvm: program not executable

stack traceback:
        [C]: in function 'saveobj'
        C:\Users\zdevito\terra\src\terralib.lua:3979: in function 'saveobj'
        ..\share\terra\tests\hello.t:8: in main chunk.

I found out change executable to .o, will let hello.t execute without
error. But the .o still could not be used because of the undefined external
error.

I don't know where I can link to that legacy version. I tried
terralib.linklibrary
or add them to environment variable, both not work. So I'm wondering if you
know where I should link to that .lib. Thank you.

Sincerely,

Yan
